src/api/weth/WethApi.ts

```diff
@@ -0,0 +1,116 @@
+import Web3 from 'web3'
+
+import { Network, WithTxOptionalParams, Receipt } from 'types'
+import { WETH_ADDRESS_MAINNET, WETH_ADDRESS_RINKEBY } from 'const'
+import { wethAbi } from '@gnosis.pm/dex-js'
+import { logDebug } from 'utils'
+
+import { WethContract } from '@gnosis.pm/dex-js/build-esm/contracts/WethContract'
+
+export interface WrapUnwrapParams extends WithTxOptionalParams {
+  networkId: number
+  userAddress: string
+  amount: string
+}
+
+/**
+ * Interfaces for Wrapping ETH and unwrapping WETH token
+ */
+export interface WethApi {
+  deposit(params: WrapUnwrapParams): Promise<Receipt>
+  withdraw(params: WrapUnwrapParams): Promise<Receipt>
+}
+
+export interface WethApiDependencies {
+  web3: Web3
+  fetchGasPrice(): Promise<string | undefined>
+}
+
+function getWethAddressByNetwork(networkId: number): string {
+  switch (networkId) {
+    case Network.Mainnet:
+      return WETH_ADDRESS_MAINNET
+    case Network.Rinkeby:
+      return WETH_ADDRESS_RINKEBY
+    default:
+      throw new Error(`WethApi was not deployed to network ${networkId}`)
+  }
+}
+
+/**
+ * Basic implementation of WETH API
+ * Note that there's already another API for all ERC20, so only the deposit/withdraw is implemented in this API
+ */
+export class WethApiImpl implements WethApi {
+  protected _contractPrototype: WethContract
+  private web3: Web3
+  private fetchGasPrice: WethApiDependencies['fetchGasPrice']
+  protected static _contractsCache: { [network: number]: { [address: string]: WethContract } } = {}
+
+  public constructor(injectedDependencies: WethApiDependencies) {
+    Object.assign(this, injectedDependencies)
+    this._contractPrototype = new this.web3.eth.Contract(wethAbi) as WethContract
+  }
+
+  public async deposit(params: WrapUnwrapParams): Promise<Receipt> {
+    const { networkId, amount, userAddress, txOptionalParams } = params
+    const contract = await this._getContract(networkId)
```

Comment:
why not 
```ts
const [contract, gasPrice] = await Promise.all([
  this._getContract(networkId),
  this.fetchGasPrice(),
])
```

I think `await`s inside send tx calls are weird

Reply:
well, yes, getContract is cached, so after the first query is just sync code anyways. So I thought it's was slightly more readable than Promise.all while not big advantage